llvm/tools/sycl-post-link/SpecConstants.cpp

```diff
@@ -878,6 +902,10 @@ PreservedAnalyses SpecConstantsPass::run(Module &M,
     }
   }
 
+  if (Mode == HandlingMode::default_values)
+    for (Function *F : SpecConstDeclarations)
+      F->removeFromParent();
```

Comment:
We do, regardless of optimization level:

https://github.com/intel/llvm/blob/7697e7bf9ab528a3efe16dbc2b168db825a96b5b/llvm/tools/sycl-post-link/ModuleSplitter.cpp#L306-L319

https://github.com/intel/llvm/blob/7697e7bf9ab528a3efe16dbc2b168db825a96b5b/llvm/tools/sycl-post-link/ModuleSplitter.cpp#L554-L563

Reply:
Well, yes. However, we copy device image after this cleanup.

Reply:
If I see it correctly the current SpecConstPass leaves abandoned function declarations which are not removed later.

Reply:
I still don't think that removal of unused declarations should be a part of `SpecConstantsPass`, because _any_ transformation pass could leave those and having cleanup code in _each_ of them does not seem like a good design. We have a dedicated pass for cleanup, no need to duplicate its code to every pass.

You can just add `StripDeadPrototypesPass` to the pipeline when launching `SpecConstantsPass` to emit that extra device image with default values of spec constants

Reply:
Added StripDeadPrototypesPass